### PR TITLE
Spec Actors integration

### PR DESCRIPTION
**Pull request by magik6k**
Work in progress..

### CODE REVIEW

File: chain/gen/genesis/miners.go [hunk 1]
```diff
@@ -86,10 +90,12 @@ func SetupStorageMiners(ctx context.Context, cs *store.ChainStore, sroot cid.Cid
 		{
 			params := &market.PublishStorageDealsParams{}
 			for _, preseal := range m.Sectors {
+
 				params.Deals = append(params.Deals, market.ClientDealProposal{
 					Proposal:        preseal.Deal,
-					ClientSignature: crypto.Signature{},
+					ClientSignature: crypto.Signature{}, // TODO: do we want to sign these? Or do we want to fake signatures for genesis setup?
```
Comment (magik6k) [inline chain/gen/genesis/miners.go]: This depends on whether we want to support deals in genesis

File: chain/gen/gen.go [hunk 2]
```diff
@@ -276,8 +274,13 @@ func CarWalkFunc(nd format.Node) (out []*format.Link, err error) {
 }
 
 func (cg *ChainGen) nextBlockProof(ctx context.Context, pts *types.TipSet, m address.Address, round int64) (*types.EPostProof, *types.Ticket, error) {
+	mc := &mca{w: cg.w, sm: cg.sm}
 
-	lastTicket := pts.MinTicket()
+	// TODO: REVIEW: Am I doing this correctly?
+	ticketRand, err := mc.ChainGetRandomness(ctx, pts.Key(), crypto.DomainSeparationTag_TicketProduction, pts.Height(), m.Bytes())
```
Comment (whyrusleeping) [inline chain/gen/gen.go]: i heard some rumblings about people wanting the inputs to randomness to be a cbor object. though i'm not sure if thats supposed to be the object here, or the one we pass into computeVRF

File: chain/state/statetree.go [hunk 3]
```diff
@@ -28,14 +30,14 @@ type StateTree struct {
...
-		root:       hamt.NewNode(cst),
+		root:       hamt.NewNode(cst, hamt.UseTreeBitWidth(5)),
 		Store:      cst,
 		actorcache: make(map[address.Address]*types.Actor),
 	}, nil
 }
 
 func LoadStateTree(cst cbor.IpldStore, c cid.Cid) (*StateTree, error) {
-	nd, err := hamt.LoadNode(context.Background(), cst, c)
+	nd, err := hamt.LoadNode(context.Background(), cst, c, hamt.UseTreeBitWidth(5))
```
Comment (whyrusleeping) [inline chain/state/statetree.go]: Each of these is a "TODO: use the spec thingy"

File: chain/store/store.go [hunk 4]
```diff
@@ -855,22 +889,30 @@ func (cs *ChainStore) TryFillTipSet(ts *types.TipSet) (*FullTipSet, error) {
 	return NewFullTipSet(out), nil
 }
 
-func drawRandomness(t *types.Ticket, round int64) []byte {
-	h := sha256.New()
-	var buf [8]byte
-	binary.LittleEndian.PutUint64(buf[:], uint64(round))
-
+func drawRandomness(t *types.Ticket, pers crypto.DomainSeparationTag, round int64, entropy []byte) ([]byte, error) {
+	// TODO: Make this spec compliant
```
Comment (whyrusleeping) [inline chain/store/store.go]: Flagging this TODO

File: chain/store/store.go [hunk 5]
```diff
@@ -855,22 +889,30 @@ func (cs *ChainStore) TryFillTipSet(ts *types.TipSet) (*FullTipSet, error) {
...
 }
 
-func (cs *ChainStore) GetRandomness(ctx context.Context, blks []cid.Cid, round int64) ([]byte, error) {
+func (cs *ChainStore) GetRandomness(ctx context.Context, blks []cid.Cid, pers crypto.DomainSeparationTag, round int64, entropy []byte) (out []byte, err error) {
 	_, span := trace.StartSpan(ctx, "store.GetRandomness")
 	defer span.End()
 	span.AddAttributes(trace.Int64Attribute("round", round))
 
+	defer func() {
+		log.Warnf("getRand %v %d %d %x -> %x", blks, pers, round, entropy, out)
```
Comment (whyrusleeping) [inline chain/store/store.go]: a warn here may be excessive, but whatever

File: chain/store/store.go [hunk 6]
```diff
@@ -855,22 +889,30 @@ func (cs *ChainStore) TryFillTipSet(ts *types.TipSet) (*FullTipSet, error) {
...
+		return nil, xerrors.Errorf("deriving randomness: %w", err)
+	}
+	h.Write(entropy)
 
-	return h.Sum(nil)
+	return h.Sum(nil), nil
 }
 
-func (cs *ChainStore) GetRandomness(ctx context.Context, blks []cid.Cid, round int64) ([]byte, error) {
+func (cs *ChainStore) GetRandomness(ctx context.Context, blks []cid.Cid, pers crypto.DomainSeparationTag, round int64, entropy []byte) (out []byte, err error) {
```
Comment (whyrusleeping) [inline chain/store/store.go]: cc @sternhenri for review

File: chain/types/ask.go [hunk 7]
```diff
@@ -1,27 +1,9 @@
...
-	// Price per GiB / Epoch
-	Price BigInt
-
-	MinPieceSize uint64
-	Miner        address.Address
-	Timestamp    uint64
-	Expiry       uint64
-	SeqNo        uint64
-}
+type StorageAsk = storagemarket.StorageAsk
```
Comment (whyrusleeping) [inline chain/types/ask.go]: we should make a note to just squash all these aliased types into using the spec ones. Def not needed now, but easy to do over time

File: chain/types/bigint.go [hunk 8]
```diff
@@ -1,41 +1,21 @@
...
-	"golang.org/x/xerrors"
+	"github.com/filecoin-project/lotus/build"
 )
 
 const BigIntMaxSerializedLen = 128 // is this big enough? or too big?
 
 var TotalFilecoinInt = FromFil(build.TotalFilecoin)
 
-func init() {
-	cbor.RegisterCborType(atlas.BuildEntry(BigInt{}).Transform().
```
Comment (whyrusleeping) [inline chain/types/bigint.go]: :pray:

File: chain/vm/invoker.go [hunk 9]
```diff
@@ -33,18 +40,19 @@ func NewInvoker() *invoker {
 	}
 
 	// add builtInCode using: register(cid, singleton)
+	inv.Register(builtin.SystemActorCodeID, system.Actor{}, adt.EmptyValue{})
 	inv.Register(actors.InitCodeCid, actors.InitActor{}, actors.InitActorState{})
```
Comment (whyrusleeping) [inline chain/vm/invoker.go]: TODO: directly use the specs actors here

File: chain/vm/spec_shim.go [hunk 10]
```diff
@@ -3,36 +3,64 @@ package vm
...
 	defer func() {
 		if r := recover(); r != nil {
 			if ar, ok := r.(aerrors.ActorError); ok {
+				fmt.Fprintln(os.Stderr, "VM.Call failure: ", ar)
+				debug.PrintStack()
 				aerr = ar
 				return
 			}
-			fmt.Println("caught one of those actor errors: ", r)
+			fmt.Fprintln(os.Stderr, "caught one of those actor errors: ", r)
```
Comment (whyrusleeping) [inline chain/vm/spec_shim.go]: we could log.Warn these I guess

File: chain/vm/spec_shim.go [hunk 11]
```diff
@@ -56,62 +84,156 @@ func (rs *runtimeShim) shimCall(f func() interface{}) (rval []byte, aerr aerrors
...
-	panic(aerrors.Newf(uint8(code), msg, args...))
+func (rs *runtimeShim) Store() vmr.Store {
+	return rs
 }
 
-func (rs *runtimeShim) AbortStateMsg(msg string) {
-	rs.Abort(101, msg)
+func (rs *runtimeShim) NewActorAddress() address.Address {
+	var b bytes.Buffer
+	if err := rs.ImmediateCaller().MarshalCBOR(&b); err != nil { // todo: spec says cbor; why not just bytes?
```
Comment (whyrusleeping) [inline chain/vm/spec_shim.go]: Yeah, seems odd...

File: chain/vm/spec_shim.go [hunk 12]
```diff
@@ -56,62 +84,156 @@ func (rs *runtimeShim) shimCall(f func() interface{}) (rval []byte, aerr aerrors
...
+		rs.Abortf(exitcode.ErrSerialization, "writing caller address into a buffer: %v", err)
+	}
+
+	var err error
```
Comment (whyrusleeping) [inline chain/vm/spec_shim.go]: this declaration can be dropped

File: chain/vm/syscalls.go [hunk 13]
```diff
@@ -1,69 +1,131 @@
...
-		},
+func Syscalls(verifier sectorbuilder.Verifier) runtime.Syscalls {
+	return &syscallShim{verifier}
+}
+
+type syscallShim struct {
+	verifier sectorbuilder.Verifier
+}
+
+func (ss *syscallShim) ComputeUnsealedSectorCID(ssize abi.SectorSize, pieces []abi.PieceInfo) (cid.Cid, error) {
```
Comment (whyrusleeping) [inline chain/vm/syscalls.go]: TODO: build a big table test for this function

File: chain/vm/vm.go [hunk 14]
```diff
@@ -543,6 +563,37 @@ func (vm *VM) Flush(ctx context.Context) (cid.Cid, error) {
 	return root, nil
 }
 
+// vm.MutateState(idAddr, func(cst cbor.IpldStore, st *ActorStateType) error {...})
+func (vm *VM) MutateState(ctx context.Context, addr address.Address, fn interface{}) error {
```
Comment (whyrusleeping) [inline chain/vm/vm.go]: mmmmmmmmmmmmmmmmmmmmmmmmmmm, ok

File: lib/zerocomm/zerocomm.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+	{0x7d, 0xf2, 0x9b, 0x69, 0x77, 0x31, 0x99, 0xe8, 0xf2, 0xb4, 0xb, 0x77, 0x91, 0x9d, 0x4, 0x85, 0x9, 0xee, 0xd7, 0x68, 0xe2, 0xc7, 0x29, 0x7b, 0x1f, 0x14, 0x37, 0x3, 0x4f, 0xc3, 0xc6, 0x2c},
+	{0x66, 0xce, 0x5, 0xa3, 0x66, 0x75, 0x52, 0xcf, 0x45, 0xc0, 0x2b, 0xcc, 0x4e, 0x83, 0x92, 0x91, 0x9b, 0xde, 0xac, 0x35, 0xde, 0x2f, 0xf5, 0x62, 0x71, 0x84, 0x8e, 0x9f, 0x7b, 0x67, 0x51, 0x7},
+	{0xd8, 0x61, 0x2, 0x18, 0x42, 0x5a, 0xb5, 0xe9, 0x5b, 0x1c, 0xa6, 0x23, 0x9d, 0x29, 0xa2, 0xe4, 0x20, 0xd7, 0x6, 0xa9, 0x6f, 0x37, 0x3e, 0x2f, 0x9c, 0x9a, 0x91, 0xd7, 0x59, 0xd1, 0x9b, 0x1},
+	{0x6d, 0x36, 0x4b, 0x1e, 0xf8, 0x46, 0x44, 0x1a, 0x5a, 0x4a, 0x68, 0x86, 0x23, 0x14, 0xac, 0xc0, 0xa4, 0x6f, 0x1, 0x67, 0x17, 0xe5, 0x34, 0x43, 0xe8, 0x39, 0xee, 0xdf, 0x83, 0xc2, 0x85, 0x3c},
+	{0x7, 0x7e, 0x5f, 0xde, 0x35, 0xc5, 0xa, 0x93, 0x3, 0xa5, 0x50, 0x9, 0xe3, 0x49, 0x8a, 0x4e, 0xbe, 0xdf, 0xf3, 0x9c, 0x42, 0xb7, 0x10, 0xb7, 0x30, 0xd8, 0xec, 0x7a, 0xc7, 0xaf, 0xa6, 0x3e},
+	{0xe6, 0x40, 0x5, 0xa6, 0xbf, 0xe3, 0x77, 0x79, 0x53, 0xb8, 0xad, 0x6e, 0xf9, 0x3f, 0xf, 0xca, 0x10, 0x49, 0xb2, 0x4, 0x16, 0x54, 0xf2, 0xa4, 0x11, 0xf7, 0x70, 0x27, 0x99, 0xce, 0xce, 0x2},
```
Comment (whyrusleeping) [inline lib/zerocomm/zerocomm.go]: ok

File: markets/storageadapter/client.go [hunk 16]
```diff
@@ -295,35 +303,30 @@ func (c *ClientNodeAdapter) OnDealSectorCommitted(ctx context.Context, provider
...
-	}
-	signature, err := utils.ToSharedSignature(localProposal.ProposerSignature)
-	if err != nil {
-		return err
-	}
-	proposal.ProposerSignature = signature
-	return nil
+func (n *ClientNodeAdapter) SignProposal(ctx context.Context, signer address.Address, proposal samarket.DealProposal) (*samarket.ClientDealProposal, error) {
+	// TODO: output spec signed proposal
+	log.Warn("TODO: stub SignProposal")
```
Comment (whyrusleeping) [inline markets/storageadapter/client.go]: lets not forget to do these signatures...

File: node/impl/paych/paych.go [hunk 17]
```diff
@@ -103,6 +105,8 @@ func (a *PaychAPI) PaychStatus(ctx context.Context, pch address.Address) (*api.P
 }
 
 func (a *PaychAPI) PaychClose(ctx context.Context, addr address.Address) (cid.Cid, error) {
+	panic("TODO Settle logic")
```
Comment (whyrusleeping) [inline node/impl/paych/paych.go]: yes

File: storage/fpost_run.go [hunk 18]
```diff
@@ -102,45 +107,45 @@ func (s *FPoStScheduler) checkFaults(ctx context.Context, ssi sectorbuilder.Sort
 		}
 	}
 
+	var faultIDs []abi.SectorNumber
 	if len(faults) > 0 {
-		params := &actors.DeclareFaultsParams{Faults: types.NewBitField()}
+		params := &miner.DeclareTemporaryFaultsParams{
+			Duration:      900, // TODO: duration is annoying
```
Comment (whyrusleeping) [inline storage/fpost_run.go]: we should remove temp fault durations... seems fine to me

File: storage/sealing/states.go [hunk 19]
```diff
@@ -174,17 +180,18 @@ func (m *Sealing) handleWaitSeed(ctx statemachine.Context, sector SectorInfo) er
 func (m *Sealing) handleCommitting(ctx statemachine.Context, sector SectorInfo) error {
 	log.Info("scheduling seal proof computation...")
 
+	log.Infof("KOMIT %d %x(%d); %x(%d); %v; r:%x; d:%x", sector.SectorID, sector.Ticket.TicketBytes, sector.Ticket.BlockHeight, sector.Seed.TicketBytes, sector.Seed.BlockHeight, sector.pieceInfos(), sector.CommR, sector.CommD)
```
Comment (whyrusleeping) [inline storage/sealing/states.go]: :fire:

File: chain/actors/actor_cron.go [hunk 20]
```diff
@@ -1,48 +1,9 @@
...
 
 import (
-	"github.com/filecoin-project/go-address"
-	"github.com/filecoin-project/lotus/chain/actors/aerrors"
-	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/specs-actors/actors/builtin/cron"
 )
 
-type CronActor struct{}
+type CronActor = cron.Actor
```
Comment (Kubuxu) [inline chain/actors/actor_cron.go]: Can we do it without type aliases (reference spec-actors directly)?

Comment (whyrusleeping) [reply chain/actors/actor_cron.go]: yes, but i think we can just ship this now and clean things later

File: chain/gen/genesis/genesis.go [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+		return nil, xerrors.Errorf("setup init actor: %w", err)
+	}
+	if err := state.SetActor(actors.InitAddress, initact); err != nil {
+		return nil, xerrors.Errorf("set init actor: %w", err)
+	}
+
+	// Setup reward
+	err = state.SetActor(actors.RewardActor, &types.Actor{
+		Code:    builtin.RewardActorCodeID,
+		Balance: big.Int{Int: build.InitialReward},
```
Comment (Kubuxu) [inline chain/gen/genesis/genesis.go]: Should be `build.MiningRewardTotal`.

File: chain/stmgr/utils.go [hunk 22]
```diff
@@ -216,45 +164,38 @@ func GetSectorsForElectionPost(ctx context.Context, sm *StateManager, ts *types.
...
 
 func GetMinerSlashed(ctx context.Context, sm *StateManager, ts *types.TipSet, maddr address.Address) (uint64, error) {
-	var mas actors.StorageMinerActorState
-	_, err := sm.LoadActorState(ctx, maddr, &mas, ts)
-	if err != nil {
-		return 0, xerrors.Errorf("(get mslash) failed to load miner actor state: %w", err)
-	}
-
-	return mas.SlashedAt, nil
+	log.Warn("stub GetMinerSlashed")
```
Comment (Kubuxu) [inline chain/stmgr/utils.go]: //TODO

File: chain/types/message.go [hunk 23]
```diff
@@ -22,10 +23,26 @@ type Message struct {
 	GasPrice BigInt
 	GasLimit BigInt
 
-	Method uint64 // TODO: decide
+	Method abi.MethodNum
 	Params []byte
 }
 
+func (t *Message) BlockMiner() address.Address {
+	panic("implement me")
```
Comment (Kubuxu) [inline chain/types/message.go]: Flagging TODO

File: node/hello/cbor_gen.go [hunk 24]
```diff
@@ -105,7 +105,6 @@ func (t *HelloMessage) UnmarshalCBOR(r io.Reader) error {
 	if maj != cbg.MajUnsignedInt {
 		return fmt.Errorf("wrong type for uint64 field")
 	}
-	t.HeaviestTipSetHeight = uint64(extra)
```
Comment (Kubuxu) [inline node/hello/cbor_gen.go]: This generated code is weird.

Comment (Kubuxu) [reply node/hello/cbor_gen.go]: It won't write to `HeaviestTipSetHeight` ever.

File: chain/actors/actor_cron.go (hunk 20 above)
Comment (Kubuxu) [reply chain/actors/actor_cron.go]: SGWM

File: node/hello/cbor_gen.go (hunk 24 above)
Comment (magik6k) [reply node/hello/cbor_gen.go]: May need to regen, I've probably removed it when fixing compilation after merge

File: chain/vm/spec_shim.go (hunk 12 above)
Comment (magik6k) [reply chain/vm/spec_shim.go]: It's there because `rs.vmctx.StateTree` returns `ActorError`

**Comment by magik6k** [comment]
:ship:

File: chain/store/store.go [hunk 25]
```diff
@@ -855,22 +889,30 @@ func (cs *ChainStore) TryFillTipSet(ts *types.TipSet) (*FullTipSet, error) {
...
-	var buf [8]byte
-	binary.LittleEndian.PutUint64(buf[:], uint64(round))
-
+func drawRandomness(t *types.Ticket, pers crypto.DomainSeparationTag, round int64, entropy []byte) ([]byte, error) {
+	// TODO: Make this spec compliant
+	h := blake2b.New256()
+	if err := binary.Write(h, binary.BigEndian, int64(pers)); err != nil {
+		return nil, xerrors.Errorf("deriving randomness: %w", err)
+	}
 	h.Write(t.VRFProof)
```
Comment (sternhenri) [inline chain/store/store.go]: We want to use a VRFDigest here, which is to say H(VRFProof) (sorry, in a different VRF you couldn't use the Proof so it's a useless hash for now, but the right interface).

File: chain/store/store.go [hunk 26]
```diff
@@ -880,26 +922,29 @@ func (cs *ChainStore) GetRandomness(ctx context.Context, blks []cid.Cid, round i
...
 			// round is negative
-			thash := drawRandomness(mtb.Ticket, round*-1)
+			thash, err := drawRandomness(mtb.Ticket, pers, round*-1, entropy)
+			if err != nil {
+				return nil, err
+			}
 
 			// for negative lookbacks, just use the hash of the positive tickethash value
-			h := sha256.Sum256(thash)
+			h := blake2b.Sum256(thash)
```
Comment (sternhenri) [inline chain/store/store.go]: None of this is necessary anymore.

File: chain/store/store.go [hunk 27]
```diff
@@ -880,26 +922,29 @@ func (cs *ChainStore) GetRandomness(ctx context.Context, blks []cid.Cid, round i
 		mtb := nts.MinTicketBlock()
 
 		if int64(nts.Height()) <= round {
```
Comment (sternhenri) [inline chain/store/store.go]: try:
` 		if int64(nts.Height()) <= round  || mtb.Height == 0`
instead. And remove the below special case.

**Pull request merged**